### PR TITLE
Fixes #2656 too many parameters

### LINKED ISSUE #2656: Too many parameters created for Individual
There are too many parameters added to individual building blocks when converting to a core building block.

Individual building block should only include species specific parameters.

### PR DESCRIPTION
**Pull request by msevestre**

### CODE REVIEW

**Comment by msevestre** [comment]
@rwmcintosh @Yuri05 This is a work in progress because of #2705

File: src/PKSim.Assets/PKSimConstants.cs [hunk 1]
```diff
@@ -606,13 +606,6 @@ public static string ProjectFileVersion4IsNotSupportedAnymore(string projectFile
 
          public static string UnableToUpdateParameterException(string parameterPath, string simulationName) => $"Unable to update parameter.\nParameter with path '{parameterPath}' not found in simulation '{simulationName}'.";
 
-         public static string ConstantParameterAlreadyExistsInContainer(string containerName, string parameterName) => $"Parameter '{parameterName}' already exists in '{containerName}' but is defined as a constant parameter.";
```
Comment (msevestre) [inline src/PKSim.Assets/PKSimConstants.cs:609]: removes all of those constants that were just developer errors but we don't do errors anymore. Now we throw exception

File: src/PKSim.Core/Mappers/IndividualToIndividualBuildingBlockMapper.cs [hunk 2]
```diff
@@ -60,7 +64,17 @@ protected override IFormula TemplateFormulaFor(IParameter parameter, IFormulaCac
...
       protected override IReadOnlyList<IParameter> AllParametersFor(Individual individual)
       {
-         return individual.GetAllChildren<IParameter>().Where(x => x.GroupName != CoreConstants.Groups.RELATIVE_EXPRESSION).ToList();
+         bool shouldExportParameter(IParameter parameter)
+         {
+            //these parameters are exported separately
+            if (parameter.GroupName == RELATIVE_EXPRESSION)
+               return false;
+
+            return !_sameFormulaOrValueForAllSpeciesRepository.IsSameFormulaOrValue(parameter);
```
Comment (msevestre) [inline src/PKSim.Core/Mappers/IndividualToIndividualBuildingBlockMapper.cs:73]: my idea is that I am only going to export the parameter in the individual BB if it's not the same for all species (value or formula). Seems reasonable?

File: src/PKSim.Core/Model/IndividualParameterSameFormulaOrValueForAllSpecies.cs [hunk 3]
```diff
@@ -11,5 +7,6 @@ public class IndividualParameterSameFormulaOrValueForAllSpecies
       public string ParameterName { get; set; }
       public bool IsSameFormula { get; set; }
       public bool IsSameValue => !IsSameFormula;
+      public bool IsSame => IsSameFormula || IsSameValue;
```
Comment (msevestre) [inline src/PKSim.Core/Model/IndividualParameterSameFormulaOrValueForAllSpecies.cs:10]: often I just want to know is it the same value or formula. I don't really care for the specific. If I export, it will be exported based on the constant vs formula anyways

File: src/PKSim.Core/Model/PKSimSpatialStructureFactory.cs [hunk 4]
```diff
@@ -112,13 +112,13 @@ private void addNeighborhoods(SpatialStructure spatialStructure, Organism organi
       private void addNeighborhood(NeighborhoodBuilder neighborhood, SpatialStructure spatialStructure, OriginData originData, ModelProperties modelProperties, IFormulaCache formulaCache)
       {
          spatialStructure.AddNeighborhood(neighborhood);
-         _parameterContainerTask.AddModelParametersTo(neighborhood, originData, modelProperties, formulaCache);
-         _parameterContainerTask.AddModelParametersTo(neighborhood.MoleculeProperties, originData, modelProperties, formulaCache);
+         _parameterContainerTask.AddParametersToSpatialStructureContainer(neighborhood, originData, modelProperties, formulaCache);
```
Comment (msevestre) [inline src/PKSim.Core/Model/PKSimSpatialStructureFactory.cs:115]: REnamed thsi because it's only used when creating the spatial structure

File: src/PKSim.Core/Repositories/IIndividualParameterBySpeciesRepository.cs [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 namespace PKSim.Core.Repositories
 {
-   public interface IIndividualParametersNotCommonForAllSpeciesRepository : IStartableRepository<IndividualParameterBySpecies>
```
Comment (msevestre) [inline src/PKSim.Core/Repositories/IIndividualParameterBySpeciesRepository.cs:6]: that name was a bit hard to follow :)

File: src/PKSim.Core/Repositories/IIndividualParametersSameFormulaOrValueForAllSpeciesRepository.cs [hunk 6]
```diff
@@ -7,6 +8,7 @@ public interface IIndividualParametersSameFormulaOrValueForAllSpeciesRepository
    {
       bool IsSameFormula(ParameterMetaData parameterMetaData);
       bool IsSameValue(ParameterMetaData parameterMetaData);
-      (bool IsSameFormula, bool IsSameValue) IsSameFormulaOrValue(ParameterMetaData parameterMetaData);
+      (bool isSame, bool exists) IsSameFormulaOrValue(ParameterMetaData parameterMetaData);
```
Comment (msevestre) [inline src/PKSim.Core/Repositories/IIndividualParametersSameFormulaOrValueForAllSpeciesRepository.cs:11]: while I don't care about Formula or value, I need to know if this exists or not. false, false is not enough

File: src/PKSim.Core/Services/ParameterContainerTask.cs [hunk 7]
```diff
@@ -51,30 +56,38 @@ public ParameterContainerTask(IParameterQuery parameterQuery, IParameterFactory
...
-      public void AddModelParametersTo<TContainer>(TContainer parameterContainer, OriginData originData, ModelProperties modelProperties, IFormulaCache formulaCache) where TContainer : IContainer
+      public void AddParametersToSpatialStructureContainer<TContainer>(TContainer parameterContainer, OriginData originData, ModelProperties modelProperties, IFormulaCache formulaCache) where TContainer : IContainer
       {
-         bool addParameter(ParameterMetaData param)
+         bool addParameter(ParameterMetaData parameterMetaData)
          {
-            return param.BuildingBlockType != PKSimBuildingBlockType.Individual || _commonParametersRepository.UsedForAllSpecies(param);
+            //all non individual parameters are added to the spatial structure by default
+            //otherwise, we add all parameters that are used by all species. The value used will be set based on the fact that the parameter has a shared value for all parameters or species or not
+            return parameterMetaData.BuildingBlockType != PKSimBuildingBlockType.Individual || _individualParameterBySpeciesRepository.UsedForAllSpecies(parameterMetaData);
```
Comment (msevestre) [inline src/PKSim.Core/Services/ParameterContainerTask.cs:71]: @Yuri05 @rwmcintosh 
I think this is it. We add a parmaeter to the spt structure if it's either not an individual parameter OR if it's used by all species? This calls returns true for distributed parameters

Maybe this is right that it returns true. But this makes thinks very complicated then because we have different parameter structure afterwards (distributed vs non distrebitued)

File: src/PKSim.Core/Services/ParameterContainerTask.cs [hunk 8]
```diff
@@ -51,30 +56,38 @@ public ParameterContainerTask(IParameterQuery parameterQuery, IParameterFactory
...
-      {
-         addParametersTo(process, null, CoreConstants.CalculationMethod.ForProcesses);
+         void parameterValueModifier(ParameterMetaData parameterMetaData, IParameter parameter)
+         {
+            //here we know that the parameter fulfills the criteria above. However, in case of a parameter that is used by all species, we need to verify that either the formula or the value is common 
+            //to all species. Otherwise, we set the value to NaN
+            var (isSame, exists) = _sameFormulaOrValueForAllSpeciesRepository.IsSameFormulaOrValue(parameterMetaData);
+            if (exists && !isSame)
+            {
+               parameter.Formula = new ConstantFormula(double.NaN);
```
Comment (msevestre) [inline src/PKSim.Core/Services/ParameterContainerTask.cs:82]: I believe isSameFormulaOrValue returns TRUE for distributed parmaeter? Is this possible? I need to check

File: src/PKSim.Core/Services/ParameterContainerTask.cs [hunk 9]
```diff
@@ -115,97 +128,90 @@ public void AddFormulationParametersTo(Formulation formulation)
...
       public void AddMoleculeParametersTo(MoleculeBuilder molecule, IFormulaCache formulaCache)
-      {
-         addParametersTo(molecule, null, CoreConstants.CalculationMethod.ForCompounds, x => true, formulaCache);
-      }
+         => addParametersTo(molecule, calculationMethods: CoreConstants.CalculationMethod.ForCompounds, formulaCache: formulaCache);
 
       public void AddApplicationParametersTo(IContainer container)
       {
-         addParametersTo(container, null, CoreConstants.CalculationMethod.ForApplications);
+         addParametersTo(container, calculationMethods: CoreConstants.CalculationMethod.ForApplications);
```
Comment (msevestre) [inline src/PKSim.Core/Services/ParameterContainerTask.cs:140]: the rest in this file is just simplification of code

File: src/PKSim.Infrastructure/ORM/Repositories/IndividualParametersSameFormulaOrValueForAllSpeciesRepository.cs [hunk 10]
```diff
@@ -41,8 +51,8 @@ protected override void DoStart()
                IsSameFormula = isSameFormula
             };
             _individualParametersSameFormulaOrValue.Add(individualParameterSameFormulaOrValueForAllSpecies);
-            _allByContainerIdAndParameterName.Add((containerId, parameterName),
-               individualParameterSameFormulaOrValueForAllSpecies);
+            _allByContainerIdAndParameterName.Add((containerId, parameterName), individualParameterSameFormulaOrValueForAllSpecies);
+            _allByParameterPath.Add(new[] {containerPath, parameterName}.ToPathString(), individualParameterSameFormulaOrValueForAllSpecies);
```
Comment (msevestre) [inline src/PKSim.Infrastructure/ORM/Repositories/IndividualParametersSameFormulaOrValueForAllSpeciesRepository.cs:55]: I need to be able to query for a PARAMETER instance where I don't have the meta data anymore.

File: src/PKSim.Core/Mappers/IndividualToIndividualBuildingBlockMapper.cs (hunk 2 above)
Comment (Yuri05) [reply src/PKSim.Core/Mappers/IndividualToIndividualBuildingBlockMapper.cs:73]: yes

File: src/PKSim.Core/Services/ParameterContainerTask.cs (hunk 7 above)
Comment (Yuri05) [reply src/PKSim.Core/Services/ParameterContainerTask.cs:71]: Is solved now (distributed parameters not returned as common for all species anymore)

File: src/PKSim.Core/Services/ParameterContainerTask.cs (hunk 8 above)
Comment (Yuri05) [reply src/PKSim.Core/Services/ParameterContainerTask.cs:82]: same

**Review by Yuri05 — APPROVED** [review]

**Comment by msevestre** [comment]
@Yuri05 Looking at this again. It's very complicated because now we have the following
Some parameters are defined as Individual parameters but are actually now in the spatial structure. So when we create the simulation....we are missing the link to the individual parameter of origin 

image

This is a debug view in PKSim. All the parameters of type individual should have a BBId and a ParmaterId( e.g. where are they coming from). That means I need to add some of this somehow back...

**Comment by msevestre** [comment]
@Yuri05 Ok I think i finally have it...

File: src/PKSim.Core/Services/SimulationModelCreator.cs [hunk 11]
```diff
@@ -66,20 +66,27 @@ public void CreateModelFor(Simulation simulation, bool shouldValidate = true, bo
...
          var individual = simulation.Individual;
 
+         var allIndividualParameters = _containerTask.CacheAllChildren<IParameter>(individual);
+         var allSimulationParameters = _containerTask.CacheAllChildren<IParameter>(simulation.Model.Root);
+
+         //Once the model has been created, it is necessary to update origin to ensure that references are set properly
+         //Simulation id for all parameters
+         _parameterIdUpdater.UpdateSimulationId(simulation);
+         //Parameter Ids for all parameters coming from individuals (as some where created in the spatial structure independently from the individual)
+         _parameterIdUpdater.UpdateParameterIds(allIndividualParameters, allSimulationParameters);
```
Comment (msevestre) [inline src/PKSim.Core/Services/SimulationModelCreator.cs:79]: @Yuri05 This is now required to ensure that we still have the connection individual <=> simulation in PKsim.
